---
title: "Analysis Approach 1"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document: default
---

First a number of required packages are loaded and the haploplotR package is installed.
```{r Setting up the environment, eval=FALSE, include=T}
.rs.restartR()
library(devtools)
library(GenomicRanges)
library(grid)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(SNPlocs.Hsapiens.dbSNP144.GRCh37)
library(Homo.sapiens)
library(doMC)
library(plyr)
registerDoMC(cores = 4) # The number of cpu-cores used in the analysis

# devtools::install_github(repo = "joe-nas/haploplotR")

devtools::build()
devtools::load_all()
```


Next information about SNPs is gathered from dbSNP144 and stored as lead_snps_gr. Then, 650Kb intervals are created around the lead_SNPs_gr creating analysis_intervals_gr. Furthermore, both objects are modified to have a common format corresponding to UCSC standart.
```{r Loading lead_snps, eval=FALSE, include=T}
data(analysis1)
print(summary(lead_snps))
head(lead_snps)
print(summary(populations))
head(populations)
data("res_analysis1")
# This is only needed if you are doing the analysis from scratch, e.g. if you start with raw SNP ids. Here such data 
# comes with the data-package provided

# lead_snps_gr <- GRanges(snpsById(SNPlocs.Hsapiens.dbSNP144.GRCh37, lead_snps))
# analysis_intervals_gr <- resize(lead_snps_gr, fix = 'center', width = 6.5e5)
# seqlevelsStyle(lead_snps_gr) <- "UCSC"
# seqlevelsStyle(analysis_intervals_gr) <- "UCSC"
# genome(lead_snps_gr) <- NA
# genome(analysis_intervals_gr) <- NA
```

Given lead_snps_gr and analysis_intervals_gr we can now import 1000Genomes haplotype data to calculate linkage disequilibrium from lead-SNPs to the 1000Genomes variants. Here we use the haploplotR package to import 1000Genomes for data. First I'll show the pipeline with limited genomic locations and populations followed by an analysis of multiple genomic locations and lastly the complete analysis (commented out for practical reasons). The latter run will recreate the actual results.

#### One genomic location 3 populations

```{r Analysis1 for a single location and few populations, eval=FALSE, include=T}
# This is the folder where 1000G data lives
ldabase <- LDABase$new(file_path = "/home/SSD-Data/1000Genomes/")

# This sets up an object specifying the analysis interval and which populations are to be analyzed
ldaimport <- LDAImport$new(ldabase = ldabase, granges =  analysis_intervals_gr[1], populations = c("CEU","CHB","JPT"))

# This imports the data
ldaimport$set_data()

# Here the imported data is associated with the analysis1 pipeline
res1 <- LDAanalysis1$new(lead_snps = lead_snps, cutoff = 0.8, lda_import = ldaimport)

## NEEDS TO BE DEBUGGED ... does not work with only one analysis interval
res1$set_rsquared()

```

#### Two genomic location 3 populations

```{R analysis1 with 2 analysis intervals, eval=FALSE, include=T}
# This is the folder where 1000G data lives
ldabase <- LDABase$new(file_path = "/home/SSD-Data/1000Genomes/")

# This sets up an object specifying the analysis interval and which populations are to be analyzed
res2 <- llply(analysis_intervals_gr, function(x){
  dat <- LDAImport$new(ldabase = ldabase, granges =  x, populations = populations)$set_data()
  res <- LDAanalysis1$new(lead_snps = lead_snps, cutoff = 0.8, lda_import = dat)
  res$set_rsquared()
  res$set_results()
  invisible(res)
}, .parallel = F,.inform = T)

# this collects tag snp intervals from multiple regions into one object
high_ld2 <- Reduce(c,Map(function(x) x$results, res2))
```


#### Complete analysis: 91 genomic regions and, 26 populations

```{R the complete analysis, eval=FALSE, include=T}

# This is the folder where 1000G data lives
ldabase <- LDABase$new(file_path = "/scratch/jfalck/1000G/")
print(populations)
print(lead_snps)
print(analysis_intervals_gr)

## WARNING: this should better not be executed on a desktop PC unless you have enough memory.
registerDoMC(cores = 14)
# This sets up an object specifying the analysis interval and which populations are to be analyzed
res_analysis1 <- llply(analysis_intervals_gr, function(x){
  dat <- LDAImport$new(ldabase = ldabase, granges =  x, populations = populations)$set_data()
  res <- LDAanalysis1$new(lead_snps = lead_snps, cutoff = 0.8, lda_import = dat)
  res$set_rsquared()
  res$set_results()
  invisible(res)
}, .parallel = F)
# 
names(res_analysis1) <- with(analysis_intervals_gr, paste(seqnames,start,end,sep = ":"))
high_ld_analysis1 <- Reduce(c,Map(function(x) x$results, res_complete))
save(res_analysis1, high_ld_analysis1, file="data/res_analysis1.RData")
```
Now that we have the tag-SNP interval, we just have to overlap them with p63 BS positions.
```{r}
data("res_analysis1")

p63_tagSNP_ol <- findOverlaps(p63BS, high_ld_analysis1)
p63ol <- p63BS[queryHits(p63_tagSNP_ol)]
tagsnpol <- high_ld_analysis1[subjectHits(p63_tagSNP_ol)]
mcols(p63ol) <- data.frame(p63ID=mcols(p63ol),as.data.frame(tagsnpol))
p63oltagsnp <- p63ol
save(p63oltagsnp, file="data/p63_ol_tagSNP_a1.RData")


data(p63_ol_tagSNP_a1)
```

#### Analysis 2, target genes

```{r}
library(ChIPpeakAnno)
library(biomaRt)
data(TSS.human.GRCh37)
ig <- c(read.table("/home/SSD-Data/Projects/thesis/section_appendix/tables/interestinggenes.csv", header = T, stringsAsFactors = F))
mart = useMart(biomart = "ENSEMBL_MART_ENSEMBL", host="grch37.ensembl.org", path="/biomart/martservice", dataset = "hsapiens_gene_ensembl")
ig_tss <- getBM(mart, attributes = c("chromosome_name","strand","transcription_start_site","external_gene_name","ensembl_gene_id"), values = ig$Promoter, mart = mart, filter="external_gene_name")
ig_tss_gr <- with(ig_tss, GRanges(Rle(chromosome_name),IRanges(start=transcription_start_site, end = transcription_start_site),strand = strand),names = external_gene_name)
elementMetadata(ig_tss_gr) <- data.frame(ensembl_gene_id = ig_tss$ensembl_gene_id)
ig_tss_promoters_gr <- promoters(ig_tss_gr)


## 
ig_merged_prmoters_gr <- reduce(ig_tss_promoters_gr)
## resize promoter regions to 650kb
promoter650k <-  reduce(resize(ig_merged_prmoters_gr,width = 650000,fix = 'center'))
## change seqlevelstyle to UCSC
seqlevelsStyle(p63BS) <- "NCBI"
data('analysis1')
## find p63BS within regions
p63BS_in_promoterwindows <- subsetByOverlaps(p63BS, promoter650k)

snps_ig_p63BS_promoters <- snpsByOverlaps(SNPlocs.Hsapiens.dbSNP144.GRCh37, p63BS_in_promoterwindows)
snps_a2 <- snps_ig_p63BS_promoters$RefSNP_id
analysis_intervals_a2 <- promoter650k
populations
save(snps_a2, analysis_intervals_a2, populations, file = "data/analysis2.RData")


```
#### Analysis 2 workflow
```{r }
data('analysis2')
registerDoMC(cores = 14)
ldabase <- LDABase$new(file_path = "/scratch/jfalck/1000G/")
seqlevelsStyle(analysis_intervals_a2) <- "UCSC"
res_analysis2 <- llply(analysis_intervals_a2, function(x){
  dat <- LDAImport$new(ldabase = ldabase, granges =  x, populations = populations)$set_data()
  res <- LDAanalysis1$new(lead_snps = snps_a2, cutoff = 0.8, lda_import = dat)
  res$set_rsquared()
  res$set_results()
  invisible(res)
}, .parallel = F)


```


