---
title: "Analysis Approach 1"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document: default
---

Setting up the analysis dependencies are loaded and SNP and populations to be analysed are defined. Furthermore, paths to 1000Genomes vcf and tbi files are set, as well as to the panel file.


```{r Setting up the environment, echo=TRUE, message=FALSE, warning=FALSE, paged.print=FALSE}
library(devtools)
library(GenomicRanges)
library(grid)using
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(SNPlocs.Hsapiens.dbSNP144.GRCh37)
library(Homo.sapiens)


lead_snps <- read.table("~/../Storage/Documents/Thesis1/section_results/tables/lead_snps.csv", header = T, stringsAsFactors = F)$lead.SNPs[1:5]
populations <- read.table("~/../Storage/Documents/Thesis1/section_appendix/tables/1000Genomes_populations.csv", header = T, stringsAsFactors = F, sep = ",")$Code

vcffile <- "/home/SSD-Data/1000Genomes/ALL.%s.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz"
panelfile <- "/home/SSD-Data/1000Genomes/integrated_call_samples_v3.20130502.ALL.panel"
```

Next information about SNPs is gathered from dbSNP144 and stored as lead_snps_gr. Then, 650Kb intervals are created around the lead_SNPs_gr creating analysis_intervals_gr. Furthermore, both objects are modified to have a common format corresponding to UCSC standart.

```{r}
lead_snps_gr <- GRanges(snpsById(SNPlocs.Hsapiens.dbSNP144.GRCh37, lead_snps))
analysis_intervals_gr <- resize(lead_snps_gr, fix = 'center', width = 6.5e5)
seqlevelsStyle(lead_snps_gr) <- "UCSC"
seqlevelsStyle(analysis_intervals_gr) <- "UCSC"
genome(lead_snps_gr) <- NA
genome(analysis_intervals_gr) <- NA
```

Given lead_snps_gr and analysis_intervals_gr we can now import 1000Genomes haplotype data to calculate linkage disequilibrium from lead-SNPs to the 1000Genomes variants. Here we use the haploplotR package to import 1000Genomes for data.

```{r Import data}
devtools::load_all('/home/SSD-Data/Projects/haploplotR/')
registerDoMC(cores = 2)
TGData <- llply(analysis_intervals_gr, function(x){
  gr <- x # location where rsuared is calculated
  data <- import1000GData(where = x, which = populations, vcf_file = vcffile, panel_file = panelfile) # 1000G data
  rsquared <- llply(data, function(dat) calculateLD(lead_snps,dat$genotype)) # rs for given population and lead_snps
  list(gr = gr, data = data, rsquared = rsquared) # return all the data as list
}, .parallel = TRUE)

```
Given the imported data we can now use the identifyHighLD function to find SNPs in high LD (R^2) with the lead-SNPs and to identify tag-SNP intervals.

```{R Compute R^2 and find tag-SNP intervals}

results <- llply(TGData, function(x){ ## loop over the imported 1000G data
  Reduce(c,llply(populations, function(y){ ## loop over populations, calculate LD etc and generate results.
    identifyHighLD(rsquared = x$rsquared[[y]], x$data[[y]]$info, population = y)
  }))
})

names(results) <- Reduce(c,Map(function(x) paste(x$gr,x$gr$RefSNP_id, sep = "_"), TGData))

head(results)
```

